### DISCUSSION TITLE
How can I disallow a realm admin to list other realms available on the instance?

### DISCUSSION THREAD

**Discussion by jberglinds** [keycloak/keycloak#25342] [Q&A] [answered]
I have a setup where a set of organizations need to be able to manage their own users and I'm trying to implement this using keycloak. The organizations should be isolated and not know about each others existence.

The way I've understood things, I can set this up in keycloak using realms:
* Each organization gets their own realm
* The users in each realm corresponds to the users in the organization
* For admin users, I can assign the roles from the realm-management client to allow them to manage their org by themselves. 

_**Problem is, when I log onto the realm security console as an admin user (organization admin, not master realm admin), I'm able to see other realms in the realm selector dropdown. How can I disallow this?**_ 
If I select another realm, I'm not actually able to manage it, but just the fact that I'm able to see the other realms/organizations is bad, since it leaks information about other organizations/customers in my system.

This is what it looks like when I sign into the admin for org-1 and go to the security console: (I can see the names of other realms in the dropdown)
<img width="1264" alt="Screenshot 2023-12-06 at 17 09 29" src="https://github.com/keycloak/keycloak/assets/6368728/2924b185-7a99-460b-9dd4-2be6fd0c4887">
The data is in this case being fetched from the following URL:
http://localhost:8080/admin/realms/org-1/ui-ext/realms?

As you can see, this user does not have any permissions at all. (I've emptied the default role completely):
<img width="1265" alt="Screenshot 2023-12-06 at 17 09 06" src="https://github.com/keycloak/keycloak/assets/6368728/7755ea8c-7e8c-4597-84fe-74ba1547b2f6">

Running Keycloak 23.0.1 using docker, with the "fine-grained auth" preview enabled:
```
docker run -p 8080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:23.0.1 start-dev --features="admin-fine-grained-authz"
```

What am I missing?
🙏

**Comment by antikalk** [comment, marked as answer]
imho this should be considered a bug. I am not sure, why there even is a dropdown when accessing the admin console of a realm different then master. Only when signing in from the master realm a user could have cross realm permissions anyway afaik. Which keycloak version are you using? If 23: did you already encounter this behavior with v22? 

Could you create an Issue for this one?

**Reply by jberglinds** [reply]
I was using 23.0.1. 
I just tested in 22.0 and there I'm not able to see other realms, so this was introduced in v23.

When clicking the realm selector in v22, I get a 403 when it tries to fetch the data. It seems like v23 and v22 try to get the realm information for the dropdown from different endpoints:
v22 /admin/realms
v23: /admin/realms/<realm name>/ui-ext/realms

**Reply by antikalk** [reply]
This seems to be a regression then. Will you create an issue for this bug?

**Reply by antikalk** [reply]
-> Issue created: https://github.com/keycloak/keycloak/issues/25392

**Reply by jberglinds** [reply]
Sorry, didn't get the time yesterday. Thanks for logging the issue! 🙇
